Support readable line breaks

msberends opened the pull request:
This PR makes it possible to still read line breaks in characters, which is otherwise quite annoying if you're creating e.g. plot labels with line breaks.

```r
df <- tibble(a = "test text without break", 
             b = "test text\nwith break", 
             c = "test text without break")
df
#> # A tibble: 1 x 3
#>   a                       b                    c                      
#>   <chr>                   <chr>                <chr>                  
#> 1 test text without break test text
#> with break test text without break
```

to:

```r
#> # A tibble: 1 x 3
#>   a                       b                     c                      
#>   <chr>                   <chr>                 <chr>                  
#> 1 test text without break test text\nwith break test text without break
```

The `\n` prints as the colour given in `style_na()`:
![image](https://user-images.githubusercontent.com/31037261/96848025-2c6bc200-1454-11eb-9261-fef86ddd4c03.png)

krlmlr commented:
Thanks, this is a very good point. I also like the highlighting.

I think we should escape all control characters, in the range between 1 and 31, like this:

``` r
rawToChar(as.raw(1:31))
#> [1] "\001\002\003\004\005\006\a\b\t\n\v\f\r\016\017\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037"
```

<sup>Created on 2020-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

For UTF-8, we should be able to detect them by comparing the raw character values:

``` r
charToRaw("abc\nde") < 32
#> [1] FALSE FALSE FALSE  TRUE FALSE FALSE
```

<sup>Created on 2020-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

We could do a first pass like this to see if we need replacement.

I think we should support two modes of output for character columns:

- simple: no control characters, no double quotes, no spaces at the beginning or end of the text. Simple columns can be copy-pasted into a new-style string literal of the form `r"(text)"` .
- weird: anything else. Weird columns can be used as a string literal directly.

Essentially, "weird" columns need a `deparse()` but with subtle quotes and highlighted escapes. Perhaps we can do a few `gsub()` calls on the output of `deparse()`?

msberends commented:
Great addition!

Why a check though? The transformation will be very fast. Two modes is nice, but I think the escaped newline chars (highlighted or not) instead of actually printed newline chars should always be applied to character columns. Or do I understand you incorrectly that 'simple' should have no transformation at all?

msberends commented:
I know it looks horrific, but I don't know how to *sapply or loop over those characters with `style_na("\\1")` to catch groups.

For reference:
```r

out <- gsub("\001", style_na("\\001"), out, fixed = TRUE)
out <- gsub("\002", style_na("\\002"), out, fixed = TRUE)
out <- gsub("\003", style_na("\\003"), out, fixed = TRUE)
out <- gsub("\004", style_na("\\004"), out, fixed = TRUE)
out <- gsub("\005", style_na("\\005"), out, fixed = TRUE)
out <- gsub("\006", style_na("\\006"), out, fixed = TRUE)
out <- gsub("\a", style_na("\\a"), out, fixed = TRUE)
out <- gsub("\b", style_na("\\b"), out, fixed = TRUE)
out <- gsub("\t", style_na("\\t"), out, fixed = TRUE)
out <- gsub("\n", style_na("\\n"), out, fixed = TRUE)
out <- gsub("\v", style_na("\\v"), out, fixed = TRUE)
out <- gsub("\f", style_na("\\f"), out, fixed = TRUE)
out <- gsub("\r", style_na("\\r"), out, fixed = TRUE)
out <- gsub("\016", style_na("\\016"), out, fixed = TRUE)
out <- gsub("\017", style_na("\\017"), out, fixed = TRUE)
out <- gsub("\020", style_na("\\020"), out, fixed = TRUE)
out <- gsub("\021", style_na("\\021"), out, fixed = TRUE)
out <- gsub("\022", style_na("\\022"), out, fixed = TRUE)
out <- gsub("\023", style_na("\\023"), out, fixed = TRUE)
out <- gsub("\024", style_na("\\024"), out, fixed = TRUE)
out <- gsub("\025", style_na("\\025"), out, fixed = TRUE)
out <- gsub("\026", style_na("\\026"), out, fixed = TRUE)
out <- gsub("\027", style_na("\\027"), out, fixed = TRUE)
out <- gsub("\030", style_na("\\030"), out, fixed = TRUE)
out <- gsub("\031", style_na("\\031"), out, fixed = TRUE)
out <- gsub("\032", style_na("\\032"), out, fixed = TRUE)
out <- gsub("\033", style_na("\\033"), out, fixed = TRUE)
out <- gsub("\034", style_na("\\034"), out, fixed = TRUE)
out <- gsub("\035", style_na("\\035"), out, fixed = TRUE)
out <- gsub("\036", style_na("\\036"), out, fixed = TRUE)
out <- gsub("\037", style_na("\\037"), out, fixed = TRUE)
```

krlmlr commented:
A newline is a control character. A value with an embedded newline is always weird per above definition.

I think `deparse()` will be easier:

``` r
x <- deparse("ab\001c\nde")
x <- gsub("(\\(?:0[0-3][0-7]|[abtnvfr]))", pillar::style_na("\\1"), x)
x <- gsub('(^"|"$)', pillar::style_subtle("\\1"), x)
x
#> [1] "\"ab\\001c\\nde\""
```

<sup>Created on 2020-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

To check, we can really grep for control characters too, no need to do `charToRaw()`:

``` r
cc <- "\001-\037"

rx_first <- paste0('[^" ', cc, "]")
rx_middle <- paste0('[^"', cc, "]")
rx_last <- rx_first

rx <- paste0(
  "^",
  rx_first,
  "(?:",
  "$|", rx_middle, "*", rx_last,
  ")",
  "$"
)

grep(rx, c(
  "a", "a b", " a", "a ", "a\nb", '"a"'
))
#> [1] 1 2
```

<sup>Created on 2020-10-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

msberends commented:
Awesome! 

And what could be the upside for this, except for selecting and copying?

> simple: no control characters, no double quotes, no spaces at the beginning or end of the text. Simple columns can be copy-pasted into a new-style string literal of the form r"(text)" .

I think it shouldn’t be a feature in tabular printing functions to assist in this. If users want to copy a certain value ‘asis’, I think they should use other methods. But if you want this pkg to be as comprehensive as possible, then of course you could 🙂

krlmlr commented:
I don't understand the question.

I do believe we should make copy-pasting as easy and unambiguous as possible.

msberends commented:
That answers the question, sorry for the lack of clarity.

krlmlr commented:
I was confused. This is already handled (in a slightly different way) by `utf8::utf8_encode()`. I have opened a new issue.
